presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorNodePartitioningProvider.java

```diff
@@ -52,4 +56,56 @@ BucketFunction getBucketFunction(
             ConnectorPartitioningHandle partitioningHandle,
             List<Type> partitionChannelTypes,
             int bucketCount);
+
+    class ConnectorBucketNodeMap
+    {
+        private final int bucketCount;
+        private final Optional<Map<Integer, Node>> bucketToNode;
+
+        public ConnectorBucketNodeMap(int bucketCount)
+        {
+            this(bucketCount, Optional.empty());
+        }
+
+        public ConnectorBucketNodeMap(Map<Integer, Node> bucketToNode)
+        {
+            this(
+                    bucketToNode.keySet().stream()
+                            .mapToInt(Integer::intValue)
+                            .max()
+                            .getAsInt() + 1,
+                    Optional.of(bucketToNode));
+        }
+
+        private ConnectorBucketNodeMap(int bucketCount, Optional<Map<Integer, Node>> bucketToNode)
+        {
+            if (bucketCount <= 0) {
+                throw new IllegalArgumentException("bucketCount must be positive");
+            }
+            if (bucketToNode.isPresent() && bucketToNode.get().size() != bucketCount) {
+                throw new IllegalArgumentException(format("Mismatched bucket count in bucketToNode (%s) and bucketCount (%s)", bucketToNode.get().size(), bucketCount));
+            }
+            this.bucketCount = bucketCount;
+            this.bucketToNode = requireNonNull(bucketToNode, "bucketCount is null");
```

Comment:
- `bucketCount` -> `bucketToNode` 
- this null check is after the access to the `bucketToNode` on L85.

Reply:
Switch to use static factory method.